src/libServer/Server.cpp

```diff
@@ -1376,4 +1422,46 @@ string Server::GetNodeState() {
   } else {
     return m_mediator.m_ds->GetStateString();
   }
-}
\ No newline at end of file
+}
+
+vector<uint> GenUniqueIndices(uint32_t size, uint32_t num, mt19937& eng) {
+  // case when the number required is greater than total numbers being shuffled
+  if (size < num) {
+    num = size;
+  }
+  vector<uint> v(size);
+  iota(begin(v), end(v), 0);
+  shuffle(v.begin(), v.end(), eng);
+  vector<uint> newVec(v.begin(), v.begin() + num);
+  return newVec;
+}
+
+Json::Value Server::GetShardMembers(uint32_t shardID) {
+  if (!LOOKUP_NODE_MODE) {
+    throw JsonRpcException(RPC_INVALID_REQUEST, "Sent to a non-lookup");
+  }
+  const auto& num_shards = m_mediator.m_lookup->GetShardPeers().size();
```

Comment:
I think each of the two calls to `GetShardPeers()` in this function creates a temporary anonymous `DequeOfShard` object, which can be costly. Also, the call itself needs to acquire mutex each time, and maybe in the second time the shard structure has changed (e.g., shard dropped).  Perhaps store it to a local after the first call.

Reply:
done